File: stack-medium/503.下一个更大元素-ii.java
/*
 * @lc app=leetcode.cn id=503 lang=java
 *
 * [503] 下一个更大元素 II
 */

// @lc code=start

// 224/224 cases passed (7 ms)
// Your runtime beats 92.45 % of java submissions
// Your memory usage beats 39.06 % of java submissions (40.3 MB)
class Solution {
    /* 单调栈 Monotonic stack */
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);
        // Java's stack: peek, pop, push. They all operate on the first element of the inner deque
        Deque<Integer> stack = new LinkedList<>();
        for (int i = 0; i < 2 * n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i % n]) {
                res[stack.pop()] = nums[i % n];
            }
            stack.push(i % n);
        }
        return res;
    }
}
// @lc code=end

File: stack-medium/503.下一个更大元素-ii.cpp
/*
 * @lc app=leetcode.cn id=503 lang=cpp
 *
 * [503] 下一个更大元素 II
 */

#include <stack>
#include <vector>

using namespace std;

// @lc code=start

// Wrong
class Solution {
   public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, -1);
        stack<int> s;
        // 从后往前
        for (int i = n - 1; i >= 0; i--) {
            // 单调递减栈
            while (!s.empty() && nums[i] >= s.top()) {
                s.pop();
            }
            res[i] = s.empty() ? -1 : s.top();
            s.push(nums[i]);
        }
        return res;
    }
};

// @lc code=end

// TODO: redo
// @monotonic-stack
// 224/224 cases passed (44 ms)
// Your runtime beats 58.49 % of cpp submissions
// Your memory usage beats 68.14 % of cpp submissions (22.4 MB)
class Solution_1 {
   public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> res(n, -1);
        stack<int> s;
        for (int i = 0; i < 2 * n - 1; i++) {
            while (!s.empty() && nums[s.top()] < nums[i % n]) {
                res[s.top()] = nums[i % n];
                s.pop();
            }
            s.push(i % n);
        }
        return res;
    }
};

#include <iostream>

int main() {
    auto s = Solution_1();
    // vector<int> nums{1, 2, 1};
    // vector<int> nums{5, 3, 4};
    vector<int> nums{3, -2, -1};
    auto ans = s.nextGreaterElements(nums);
    for (int i : ans) {
        cout << i << " ";
    }
}
